src/main/java/ir/sahab/nexus/plugin/tag/internal/TagStore.java

```diff
@@ -58,14 +58,22 @@ public Tag getByName(String name) {
         }
     }
 
-    public List<Tag> search(Map<String, String> attributes) {
+    /**
+     * Searches for tags with given attributes and components.
+     * @param attributes attributes to match
+     * @param componentCriteria components criteria to match on resulting tags
+     * @return list of found tags
+     */
+    public List<Tag> search(Map<String, String> attributes, List<ComponentSearchCriterion> componentCriteria) {
         try (ODatabaseDocumentTx tx = dbProvider.get().acquire()) {
             return StreamSupport.stream(entityAdapter.search(tx, attributes).spliterator(), false)
+                    .filter(tag -> tag.matches(componentCriteria))
```

Comment:
instead of loading all results to memory and then filtering them, Cant we filter them in database using SQL query?